Net shell refactoring

jukkar opened the pull request:
This PR will refactor network shell to make it nicer to work with. The current net-shell is one big (168kb) monolith file which is difficult to work with. So this PR will refactor it into smaller pieces where each net-shell command is placed into its own file. There are no functional changes in this PR, files are shuffled around and split into smaller pieces. There are lot of commits where each shell command move is in its own commit.

jukkar committed:
net: shell: Move the shell code to subsys/net/lib/shell

This is preparation commit that moves the net shell code to
subsys/net/lib/shell directory. The following commits will
then refactor the code in net_shell.c to smaller and more
manageable pieces.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Common header for all shell commands

Some useful APIs and macros for all net shell commands
to use.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add allocs command

Move "allocs" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add arp command

Move "arp" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add capture command

Move "capture" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add conn command

Move "conn" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add dns command

Move "dns" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add events command

Move "events" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add gptp command

Move "gptp" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add iface command

Move "iface" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add ipv6 command

Move "ipv6" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add ipv4 command

Move "ipv4" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add mem command

Move "mem" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add nbr command

Move "nbr" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add ping command

Move "ping" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add pkt command

Move "pkt" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add ppp command

Move "ppp" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add resume command

Move "resume" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add route command

Move "route" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add stacks command

Move "stacks" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add stats command

Move "stats" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add suspend command

Move "suspend" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add tcp command

Move "tcp" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add udp command

Move "udp" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add virtual command

Move "virtual" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add vlan command

Move "vlan" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add websocket command

Move "websocket" command to a separate file.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Removing extra stuff that is not needed

The remaining stuff that is not used in net_shell.c can be removed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Remove the stacks command

The "net stacks" has been obsolete for a long time already
so remove it for good. It is replaced by "kernel stacks" cmd.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Allow dynamic interface name expansion

Allow multiple commands use the same dynamic shell command
completion when expecting network interface index.

For example "net iface" and "net stats" are such commands.

The network interface expansion cannot be used in "net ipv6 add",
"net ipv4 add" and "net route" commands as they require more
data after the network interface index argument.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: stats: Use generic iface idx dynamic completion

The interface number is generated automatically if one
presses <tab> when expecting the interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: vlan: Enforce arg count via shell macro

Use the SHELL_CMD_ARG() to enforce the minimum parameter
count in vlan command.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Note to reviewers and thank you beforehand for all the reviews. Sorry about the amount of commits, but I wanted each net-shell command to one specific commit which then exploded the commit count. Fortunately the individual commits are just snippets taken from main net_shell.c master file. :bow:

jukkar commented:
CI is failing because some glitch in cellular modem sample.

pdgendt commented:
Maybe a dumb question, but why are the header files added?

jukkar commented:
The shell commands were created earlier with `SHELL_STATIC_SUBCMD_SET_CREATE` and I did not want to do too big changes to existing code so used a header file to set the commands. It is possible to use `SHELL_SUBCMD_SET_CREATE` in which case we could put all the commands in their own .c file.
So if you think it makes sense to get rid of those .h file, I can certainly investigate that route too.

rlubos reviewed (commented):
Nice cleanup, but indeed maybe it'd make sense to investigate whether it'd be possible to get rid of the header files.

jukkar committed:
net: shell: Rename current command as net_old

This is a preparation for the refactoring. The old "net" command
is renamed "net_old" so that the individual commands can be
placed into a separate .c files.
This is done like this because we need to use the
SHELL_SUBCMD_SET_CREATE() to create the sub-command and then
use the SHELL_SUBCMD_ADD() in the .c files to add the command
into the sub-command and not get conflict with the same name
sub-command.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
New version with the header files removed. It is much cleaner looking now. It took some time to figure out how the shell macros are to be used, the samples did not had similar usage but fortunately shell unit tests had.

rlubos reviewed (dismissed):
LGTM

cfriedt reviewed (dismissed):
LGTM - a nice follow-up commit at a later time be to conditionally compile some shell functionality based on Kconfig choices.

carlescufi commented:
@jukkar please rebase

pdgendt reviewed (dismissed):
Re-approve, assuming this is only a rebase

pdgendt commented on subsys/net/lib/shell/ping.c:
```diff
@@ -0,0 +1,502 @@
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <zephyr/random/rand32.h>
```
```suggestion
#include <zephyr/random/random.h>
```
